clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -1086,6 +1086,14 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
       Fn->setMetadata("sycl_used_aspects",
                       llvm::MDNode::get(getLLVMContext(), AspectsMD));
     }
+
+    // SYCL needs to know the source location of functions for diagnostics in
+    // the aspect propagation pass. Save the token in a srcloc metadata node.
```

Comment:
The pass that uses srcloc relies on other metadata too, so we need it to run as early as possible. As such, srcloc should also be preserved with early opts too. As you suggest in the other comment, it may make sense to add it to the tests.

Reply:
I have added a test checking that it works with early optimizations: sycl/test/warnings/aspect_perserve_srcloc.cpp